File: src/main/scala/org/jglr/phiengine/network/NetworkHandler.java
package org.jglr.phiengine.network;

import com.google.common.collect.Maps;
import org.jglr.phiengine.network.channels.NetworkChannel;
import org.jglr.phiengine.core.PhiEngine;
import org.jglr.phiengine.core.utils.Registry;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;

public class NetworkHandler {

    private final PhiEngine engine;
    private final Registry<String, NetworkChannel> channels;
    private final Registry<Class<? extends Packet>, PacketHandler<? extends Packet>> handlers;
    private final HashMap<NetworkSide, Registry<Integer, Class<? extends Packet>>> sidePackets;

    public NetworkHandler(PhiEngine engine) {
        this.engine = engine;
        channels = new Registry<>();
        handlers = new Registry<>();
        sidePackets = Maps.newHashMap();
        sidePackets.put(NetworkSide.CLIENT, new Registry<>());
        sidePackets.put(NetworkSide.SERVER, new Registry<>());
    }

    public void registerPacket(NetworkSide side, int id, Class<? extends Packet> packet) {
        sidePackets.get(side).register(id, packet);
    }

    public void registerChannel(String name, NetworkChannel channel) {
        channels.register(name, channel);
    }

    public <T extends Packet> PacketHandler<T> getHandler(Class<T> packetClass) {
        return (PacketHandler<T>) handlers.get(packetClass);
    }

    public <T extends Packet> void registerHandler(Class<T> packet, PacketHandler<T> handler) {
        handlers.register(packet, handler);
    }

    public Packet newPacket(NetworkSide side, int id) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
        Class<? extends Packet> packetClass = sidePackets.get(side).get(id);
        if(packetClass != null) {
            Packet p = packetClass.getConstructor().newInstance();
            return p;
        }
        return null;
    }

    public Server newServer() {
        Server server = new Server(this);
        return server;
    }

    public Registry<String, NetworkChannel> getChannelRegistry() {
        return channels;
    }

    public int getPacketID(NetworkSide side, Packet packet) {
        return sidePackets.get(side).findKey(packet.getClass());
    }
}

File: src/main/scala/org/jglr/phiengine/network/NetworkHandler.scala
package org.jglr.phiengine.network

import com.google.common.collect.Maps
import org.jglr.phiengine.network.NetworkSide.NetworkSide
import org.jglr.phiengine.network.channels.{PhiChannel, NetworkChannel}
import org.jglr.phiengine.core.utils.Registry
import java.lang.reflect.InvocationTargetException
import java.util

abstract class NetworkHandler {
  private final val channels: Registry[String, NetworkChannel] = new Registry[String, NetworkChannel]
  private final val handlers: Registry[Class[_ <: Packet], PacketHandler[_ <: Packet]] = new Registry[Class[_ <: Packet], PacketHandler[_ <: Packet]]
  private final val sidePackets: util.HashMap[NetworkSide, Registry[Integer, Class[_ <: Packet]]] = new util.HashMap()

  sidePackets.put(NetworkSide.CLIENT, new Registry[Integer, Class[_ <: Packet]])
  sidePackets.put(NetworkSide.SERVER, new Registry[Integer, Class[_ <: Packet]])

  def registerPacket[T <: Packet](side: NetworkSide, id: Int, packet: Class[T], handler: PacketHandler[T] = null): Unit = {
    sidePackets.get(side).register(id, packet)
    if(handler != null) {
      registerHandler(packet, handler)
    }
  }

  def registerChannel(channel: NetworkChannel) {
    channels.register(channel.getName, channel)
  }

  def getHandler[T <: Packet](packetClass: Class[T]): PacketHandler[T] = {
    handlers.get(packetClass).asInstanceOf[PacketHandler[T]]
  }

  def registerHandler[T <: Packet](packet: Class[T], handler: PacketHandler[T]) {
    handlers.register(packet, handler)
  }

  @throws(classOf[NoSuchMethodException])
  @throws(classOf[IllegalAccessException])
  @throws(classOf[InvocationTargetException])
  @throws(classOf[InstantiationException])
  def newPacket(side: NetworkSide, id: Int): Packet = {
    val packetClass: Class[_ <: Packet] = sidePackets.get(side).get(id)
    if (packetClass != null) {
      val p: Packet = packetClass.getConstructor(classOf[Int]).newInstance(id.asInstanceOf[Object])
      return p
    }
    null
  }

  def getChannelRegistry: Registry[String, NetworkChannel] = {
    channels
  }

  def getPacketID(side: NetworkSide, packet: Packet): Int = {
    val value = sidePackets.get(side).findKey(packet.getClass)
    if(value != null) {
      value
    } else {
      println("nothing found on side "+side+" for "+packet.getClass)
      -1
    }
  }
}